base/intset.jl

```diff
@@ -1,339 +1,231 @@
-function next(s::IntSet, i)
-    if i >= s.limit
-        n = Int64(i)
+@inline function in(n::Integer, s::IntSet)
+    if 1 <= n <= length(s.bits)
+        unsafe_getindex(s.bits, n)
     else
-        n = Int64(ccall(:bitvector_next, UInt64, (Ptr{UInt32}, UInt64, UInt64), s.bits, i, s.limit))
+        false
     end
-    (n, n+1)
 end
 
-isempty(s::IntSet) =
-    !s.fill1s && ccall(:bitvector_any1, UInt32, (Ptr{UInt32}, UInt64, UInt64), s.bits, 0, s.limit)==0
-
-function first(s::IntSet)
-    n = next(s,0)[1]
-    if n >= s.limit
-        throw(ArgumentError("set must be non-empty"))
-    end
-    return n
+# Use the next-set index as the state to prevent looking it up again in done
+start(s::IntSet) = next(s, 0)[2]
+function next(s::IntSet, i)
+    nextidx = i == typemax(Int) ? 0 : findnext(s.bits, i+1)
+    (i, nextidx)
 end
+done(s::IntSet, i) = i <= 0
 
-shift!(s::IntSet) = pop!(s, first(s))
 
+@noinline _throw_intset_notempty_error() = throw(ArgumentError("collection must be non-empty"))
 function last(s::IntSet)
-    if !s.fill1s
-        for i = length(s.bits):-1:1
-            w = s.bits[i]
-            if w != 0
-                return (i-1)<<5 + (31-leading_zeros(w))
-            end
-        end
-    end
-    throw(ArgumentError("set has no last element"))
-end
-
-length(s::IntSet) = Int(ccall(:bitvector_count, UInt64, (Ptr{UInt32}, UInt64, UInt64), s.bits, 0, s.limit)) +
-    (s.fill1s ? typemax(Int) - s.limit : 0)
-
-
-# Math functions
-function union!(s::IntSet, s2::IntSet)
-    if s2.limit > s.limit
-        sizehint!(s, s2.limit)
-    end
-    lim = length(s2.bits)
-    for n = 1:lim
-        s.bits[n] |= s2.bits[n]
-    end
-    if s2.fill1s
-        for n=lim+1:length(s.bits)
-            s.bits[n] = UInt32(-1)
-        end
-    end
-    s.fill1s |= s2.fill1s
-    s
+    idx = findprev(s.bits, length(s.bits))
+    idx == 0 ? _throw_intset_notempty_error() : idx
 end
 
-union(s1::IntSet) = copy(s1)
-union(s1::IntSet, s2::IntSet) = (s1.limit >= s2.limit ? union!(copy(s1), s2) : union!(copy(s2), s1))
-union(s1::IntSet, ss::IntSet...) = union(s1, union(ss...))
-
-"""
-    intersect!(s1::IntSet, s2::IntSet)
-
-Intersects sets `s1` and `s2` and overwrites the set `s1` with the result. If needed, `s1`
-will be expanded to the size of `s2`.
-"""
-function intersect!(s::IntSet, s2::IntSet)
-    if s2.limit > s.limit
-        sizehint!(s, s2.limit)
-    end
-    lim = length(s2.bits)
-    for n = 1:lim
-        s.bits[n] &= s2.bits[n]
-    end
-    if !s2.fill1s
-        for n=lim+1:length(s.bits)
-            s.bits[n] = UInt32(0)
-        end
-    end
-    s.fill1s &= s2.fill1s
-    s
-end
-
-intersect(s1::IntSet) = copy(s1)
-intersect(s1::IntSet, s2::IntSet) =
-    (s1.limit >= s2.limit ? intersect!(copy(s1), s2) : intersect!(copy(s2), s1))
-intersect(s1::IntSet, ss::IntSet...) = intersect(s1, intersect(ss...))
-
-"""
-    symdiff!(s1, s2)
+length(s::IntSet) = sum(s.bits)
```

Comment:
`countnz` might be more efficient for `BitVector`s than `sum`, but I'm not entirely sure.